### DISCUSSION TITLE
Many to Many - Multiple row creation in MariaDB

### DISCUSSION THREAD

**Discussion by rhrish** [Budibase/budibase#2779] [Help]
I've been trying to find various ways to resolve this problem but I am stuck at a dead end.

I use MariaDB with the app I am creating with BB. I have a many-to-many relationship setup like this in my DB. 

Table1 - id, name, email
Table2 - id, title, desc
Table3 - table1_id, table2_id (both ids set as foreign key references to their corresponding tables and set to restrict on delete)

I have setup a form for Table1 that does the following
- lets users enter the fields for Table1 
- lets user choose from a list of 'titles' through a multi-select picker - the data for this comes from Table2 where label is set to 'title' and value is set to 'id'. This is setup through a Data Provider.

What I am trying to achieve is this.

On clicking the 'Save' button at the end of the form
- Save row to table 1, get the ID for the newly inserted row.
- For each selection made in the multi-select picker, insert a row in the relationship table (Table 3).

How do I go about achieving this? 

I tried the 'Create Relationship' feature with the following options
- 'Many' as relationship type
- From table as 'Table1', 
- To table as 'Table2',
- Through table as 'Table3'
- From table column as 'table2'
- To table column as 'table1'

After doing this, when I went to Table1, it kept loading with the spinning wheel without any results for a long time.

Any help would be appreciated. Thanks!! 🙏🏼🙏🏼🙏🏼🙏🏼🙏🏼

**Comment by rhrish** [comment]
Additionally if there was a way to add a looping logic in automation, I could possibly loop through the list of selected IDs and insert a row in the DB.

**Comment by mjashanks** [comment]
Hey @rhrish 

Thanks for reporting this, with such a detailed description. There is most definitely something broken with many to many relationships - I have logged this here: https://github.com/Budibase/budibase/issues/2792 .

Your issue with the spinning wheel is because your "Through" table does not have an identity column. Although, fixing this won't get you any further right now.

We will need to fix this asap - it **should** work as you described above.

**Reply by rhrish** [reply]
Hello @mjashanks 

Thanks for your response. 

If it helps, I tried doing this to the tables and it helped get rid of the spinning wheel issue but was still buggy.

I changed the name of the id columns to be same across the primary key table and the foreign key table.

Table1 - **table1_id**, name, email
Table2 - **table2_id**, title, desc
Table3 - table1_id, table2_id (setup with foreign key references and restrict on delete)

I used the same method used in 'Create Relationship' as in my original post. This resolved the spinning wheel issues.

After setting up this part, I also set the 'Display Fields' for these tables - 'name' for Table1, 'title' for Table2.

This time, I went ahead and generated a form with Relationship Picker and the 'titles' field of table2 showed up in the selection list and I was able to select multiple entries.

Here is a sample of the data in the tables.

Table1
table1_id | name | email
1  | John | john@email.com
2 | Joe  | joe@email.com

Table2
table2_id| title | desc
1 | Terminator | Movie about termites
2 | Terminal | Movie about terminal
3 | Sharks | Movie about sharks

Table3 (Empty at this point)
table1_id|table2_id

Now here is where it got buggy. 

- I go to the form, enter the following details. 
- Name - John (Note that this name is the same as table1_id: 1 )
- Email - john1@email.com
- I select Terminal and Sharks from Table2

After this, the rows in Table1 show this new entry like so.

Table1
table1_id | name | email
1  | John | john@email.com
2 | Joe  | joe@email.com
**3 | John | john1@email.com**
 
However the following are the entries in Table3

Table3
table1_id|table2_id
1 | 2
1 | 3

It should have been the following
Table3
table1_id|table2_id
**3** | 2
**3** | 3

**Reply by mjashanks** [reply]
Thanks for this! it'll help us debug the issue. Going to get this fixed as soon as possible.
